core/src/main/java/org/elasticsearch/cluster/routing/AllocationId.java

```diff
@@ -148,12 +156,52 @@ public String toString() {
 
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
-        builder.startObject("allocation_id");
-        builder.field("id", id);
+        builder.startObject(ALLOCATION_ID_OBJECT_KEY);
+        builder.field(ID_KEY, id);
         if (relocationId != null) {
-            builder.field("relocation_id", relocationId);
+            builder.field(RELOCATION_ID_KEY, relocationId);
         }
         builder.endObject();
         return builder;
     }
+
+    public static AllocationId fromXContent(XContentParser parser) throws IOException {
+        XContentParser.Token token;
+        if (parser.currentToken() == null) { // fresh parser? move to the first real token under object
```

Comment:
this feels weird. I think this partially comes from the fact that the allocation id field name is serialized by this object. Instead I would change the toXContent of this one to start with startObject. Then in [ShardRouting](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java#L719) we can do:

```
if (allocationId() != null) {
            builder.field("allocation_id");
            allocationId.toXContent(builder, params);
        }
```